client/packages/programs/src/JsonForms/components/decision_tree_validation.ts

```diff
@@ -0,0 +1,94 @@
+import { DecisionNode } from './DecisionTree';
+
+// Collect node parents and collect all roots and missing nodes while doing that
+export const prepareTreeValidation = (
+  nodes: Record<string, DecisionNode>
+): {
+  roots: string[];
+  parents: Record<string, string[]>;
+  missingNodes: string[];
+} => {
+  const parents: Record<string, string[]> = {};
+  const missingNodes = [];
+  for (const key of Object.keys(nodes)) {
+    const node = nodes[key];
+    if (!node) {
+      continue;
+    }
+    // init the parents list for the current node (in case its a root node)
+    if (!parents[key]) {
+      parents[key] = [];
+    }
+    for (const child of node.branches ?? []) {
+      if (!child.node) {
+        continue;
+      }
+      if (nodes[child.node] === undefined) {
+        missingNodes.push(child.node);
+        continue;
+      }
+      const childParents = parents[child.node] ?? [];
+      if (!childParents.includes(key)) {
+        childParents.push(key);
+      }
+      parents[child.node] = childParents;
+    }
+  }
+
+  const roots = Object.entries(parents)
+    .filter(([, nodeParents]) => nodeParents.length === 0)
+    .map(([key]) => key);
+  return { roots, parents, missingNodes };
+};
+
+// Topologically sort the tree and return an error node if there is a circular dependency
```

Comment:
Would you be able to add a link (in comments) to the algorithm being used here (if it's a standard one)?

Reply:
I added one thanks